onnxruntime/core/session/inference_session.cc

```diff
@@ -2372,7 +2372,16 @@
       severity = static_cast<logging::Severity>(session_options_.session_log_severity_level);
     }
 
-    owned_session_logger_ = logging_manager_->CreateLogger(session_options_.session_logid, severity, false,
+    // There's log_id in both the Env and session_option.
+    // Inference session will use the one from Env if nothing set on session_option.
+    // Session_option can always overwrite the log id, in case the Env can be shared across sessions.
+    std::string log_id = session_options_.session_logid;    
+    std::string log_id_from_logger = logging_manager->DefaultLogger().GetLogID();
+    if (log_id.empty() && !log_id_from_logger.empty()) {
+      log_id = log_id_from_logger;
+    }
+    
```

Comment:
Do we know that this is the behavior all users want?

e.g. what if I only wanted the default logger messages to have a log id and the session ones to not? 

What's the reason the user cannot set the session options to the env log id given they created the env? That seems slightly more flexible.

Reply:
We should find the way which makes more sense. It's reasonable that the env can be shared across sessions, and the session option is more session specific. Which means the session option should have possibility to overwrite the existing settings.
Here, if there's no log id set on the session option, then the session uses the log id from the env. Otherwise if there's log id on the session option, then the session should use it from the session option.

Reply:
> Do we know that this is the behavior all users want?
> 
> e.g. what if I only wanted the default logger messages to have a log id and the session ones to not?
> 
> What's the reason the user cannot set the session options to the env log id given they created the env? That seems slightly more flexible.

I'm the author of the origin issue. In my mind, there is a hierarchy of logid's
* env is the top parent
* sessions are children of the env
* etc.
* 
In a C++ world I would do this with `<optional>` where the env is set/not, then the session set/not and each lower (more specific/deeper) in the hierarchy can override its parent. So the session can override the parent env logid by session the logid on the session. Otherwise, the log entries in all sessions within the same env has the env's logid.

The example you described would be done with pseudocode...
```
env.logid = "mycustomlogid" // sets logid for all "children" sessions part of the env
env.createsession().logid = ""  // set the logid for a specific session created as part of the env with a blank string
```

Reply:
With the current code there's no way differentiate 'not set' from 'set to empty'. Due to that, with this change the env logid would always replace an empty session logid.

My concern is that this change forces someone who relies on the existing behavior to update their code to maintain that behavior, and if they wanted an empty logid for the session there is no longer a way to do that. Conversely with the current setup you can achieve the behavior you want by explicitly setting the logid for the session. 

Of course we can pay a cost of more engineering effort to add the notion of 'optional' to the session logid, but that still doesn't counter existing users relying on the current behavior needing to update their code to maintain that. If that's say a production server-side scenario things that change how logging works can have significant impact (e.g. alerts may be setup looking for specific log message patterns).

Reply:
What's missing here is a clear spec and definition of desired behavior. Until that is clarified+agreed, the PR code should wait. Personally, I prefer to fix broken APIs and not keep errant/broken API behavior. ONNX is too young to accumulate broken APIs.

The existing api to set logid on env is broken. It can even be removed. Why? Because as in the OP issue code, the env API for logid doesn't work when it *is not* set and it doesn't work when it *is* set. Therefore, the API fails on 100% of its calls to set logid...it doesn't do anything and is irrelevant. Naturally, someone sometime spec'd it but it wasn't fully described in the spec, dev'd, or test cases written for all scenarios. Which I caught and surfaced in the issue.

There are missing test cases which should have caught this when the API was created. As part of this PR, I request a bank of test cases be created to test behavior for (not setting/set to empty string/set to full string) on both env and session in all combinations so that the decided behavor is fully tested and can catch regressions.

Writing code for nullptr, ptr to "", or ptr to something is reasonably trivial as is the hierarchy logic. But first, the behavior should be agreed. If the team wants to keep the existing broken logid behavior, then I suggest removing the api to set logid on env since it does nothing. Then it is clear that logid is only settable on the session.

Reply:
> the env API for logid doesn't work when it _is not_ set and it doesn't work when it _is_ set

The logid used in creating the Env is used for the default logger. This logger is used in places where a session logger is not available. e.g. loading execution providers into the session options which happens prior to session creation. 

https://github.com/microsoft/onnxruntime/blob/37995a7245895c1116e67a90d72c4c568dac8282/include/onnxruntime/core/common/logging/logging.h#L103

https://github.com/microsoft/onnxruntime/blob/37995a7245895c1116e67a90d72c4c568dac8282/onnxruntime/core/common/logging/logging.cc#L114

Is there something that explicitly states setting the Env logid would set the session logid? If not I wouldn't say it "does not work" or that the "api is broken". I would definitely say there's a lack of clarity about the relationship between the Env logid and the Session logid to resolve.

This could be changed to say 'The log identifier used for the default logger.  NOTE: use SetSessionLogId to set the log identifier for the session.' to at least add more clarity about the current behavior.

https://github.com/microsoft/onnxruntime/blob/37995a7245895c1116e67a90d72c4c568dac8282/include/onnxruntime/core/session/onnxruntime_c_api.h#L642-L648

I don't have a strong preference for one over the other being the future behavior. I do have a strong preference for not breaking production scenarios though, so we should understand the potential impact of changing logging output before we change a default behavior. In my experience logging is a critical component in server-side scenarios.